[inductor] make multi-kernel work with cpp-wrapper

shunting314 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #117813

Make multi-kernel work with cpp-wrapper. multi-kernel generates two equivalent variants for a reduction. At runtime the faster one is picked. But cpp-wrapper need save cubin file during codegen. They don't work with each other at the beginning.

Thanks Jason for suggesting a neat way to integrate these two. cpp-wrapper does 2 passes codegen right now. For the first pass, we still generate multi-kernel code and run it; for the second pass, we load the cubin file for the faster kernel directly. And multi-kernel python code is not generated for the second pass since they should not be needed.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

shunting314 committed:
debug multi-kernel

[ghstack-poisoned]

shunting314 committed:
Update on "debug multi-kernel"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

shunting314 committed:
Update on "debug multi-kernel"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

shunting314 committed:
Update on "debug multi-kernel"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

shunting314 committed:
Update on "debug multi-kernel"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

shunting314 committed:
Update on "[WIP] debug multi-kernel"


My previous PR cause some crash/accuracy issues when multi-kernel is enabled. This PR just enables multi-kernel by default and expose all the issue on CI. I've fixed a few and added corresponding tests. There are a few more to fix.

Overall this is still WIP. Most likely I'll disable multi-kernel by default before merging.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

jansel commented on torch/_inductor/compile_fx.py:
```diff
@@ -962,6 +962,7 @@ def compile_fx(
         with config.patch(
             {
                 "cpp_wrapper": False,
+                "triton.multi_kernel": 0,  # don't enable multi-kernel if cpp_wrapper is used
```
True?

shunting314 replied:
I checked this a bit further today. Looks like with cpp_wrapper flag is on, we do 2 passes codegen. In the second pass, we call `CudaKernelParamCache.get(kernel_name) ` to retrieve the meta data (block sizes etc) for the kernel in order to compute the grid size. However when multi-kernel is on, the CudaKernelParamCache lookup returns None for some kernels. 

I don't know the root cause yet, need look a bit more into this.

jansel reviewed (changes requested):
Is this ready to look at?  Are you looking for some comments?

shunting314 commented:
> Is this ready to look at? Are you looking for some comments?

This is just as a FYI so far. I'll need fix a few more test failures. Will notify when it's ready.

shunting314 committed:
Update on "[WIP] debug multi-kernel"


My previous PR cause some crash/accuracy issues when multi-kernel is enabled. This PR just enables multi-kernel by default and expose all the issue on CI. I've fixed a few and added corresponding tests. There are a few more to fix.

Overall this is still WIP. Most likely I'll disable multi-kernel by default before merging.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

shunting314 committed:
Update on "debug multi-kernel"


My previous PR cause some crash/accuracy issues when multi-kernel is enabled. This PR just enables multi-kernel by default and expose all the issue on CI. I've fixed a few and added corresponding tests. There are a few more to fix.

Overall this is still WIP. Most likely I'll disable multi-kernel by default before merging.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

shunting314 committed:
Update on "debug multi-kernel"


My previous PR cause some crash/accuracy issues when multi-kernel is enabled. This PR just enables multi-kernel by default and expose all the issue on CI. I've fixed a few and added corresponding tests. There are a few more to fix.

Overall this is still WIP. Most likely I'll disable multi-kernel by default before merging.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

shunting314 committed:
Update on "debug multi-kernel"


My previous PR cause some crash/accuracy issues when multi-kernel is enabled. This PR just enables multi-kernel by default and expose all the issue on CI. I've fixed a few and added corresponding tests. There are a few more to fix.

Overall this is still WIP. Most likely I'll disable multi-kernel by default before merging.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

shunting314 committed:
Update on "debug multi-kernel"


My previous PR cause some crash/accuracy issues when multi-kernel is enabled. This PR just enables multi-kernel by default and expose all the issue on CI. I've fixed a few and added corresponding tests. There are a few more to fix.

Overall this is still WIP. Most likely I'll disable multi-kernel by default before merging.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

shunting314 committed:
Update on "[inductor] make multi-kernel work with cpp-wrapper"


My previous PR cause some crash/accuracy issues when multi-kernel is enabled. This PR just enables multi-kernel by default and expose all the issue on CI. I've fixed a few and added corresponding tests. There are a few more to fix.

Overall this is still WIP. Most likely I'll disable multi-kernel by default before merging.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

shunting314 committed:
Update on "[inductor] make multi-kernel work with cpp-wrapper"


My previous PR cause some crash/accuracy issues when multi-kernel is enabled. This PR just enables multi-kernel by default and expose all the issue on CI. I've fixed a few and added corresponding tests. There are a few more to fix.

Overall this is still WIP. Most likely I'll disable multi-kernel by default before merging.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

jansel commented on torch/_inductor/codegen/multi_kernel.py:
```diff
@@ -323,6 +342,39 @@ def benchmark_sub_kernels(kernel_calls):
+        if not V.graph.record_multi_kernel_choice:
+            return
+
+        V.graph.multi_kernel_to_choice[multi_kernel_name] = choice
```
I believe the second run is a different instance of `V.graph`.  Are we caching the multi-kernel choice to disk?

If not, we should to reduce compile times.

shunting314 replied:
i think we are using the same graph object for the 2 passes of cpp-wrapper codegen: https://github.com/pytorch/pytorch/blob/main/torch/_inductor/graph.py#L1065-L1106 . But the two passes use different scheduler and wrapper objects. The comment said scheduler maybe can be reused though.

For on disk cache, yea, we do have that to reduce compilation time. The reason we don't use that cache here is that:
1. it's hard to look up that cache in `MultiKernel.run_kernel` since we need figure out the path for the cache file which is decided by the content of the python shim for the multi-kernel.
2. look up the on-disk cache is slower than just lookup an in-memory cache.

jansel reviewed (changes requested):
Tests are failing

shunting314 commented:
All tests pass on my dev machine. And the test failures on CI seems not related to PT2. Not sure why it happens. I'll ask in relevant groups tomorrow.

shunting314 committed:
Update on "[inductor] make multi-kernel work with cpp-wrapper"


Make multi-kernel work with cpp-wrapper. multi-kernel generates two equivalent variants for a reduction. At runtime the faster one is picked. But cpp-wrapper need save cubin file during codegen. They don't work with each other at the beginning.

Thanks Jason for suggesting a neat way to integrate these two. cpp-wrapper does 2 passes codegen right now. For the first pass, we still generate multi-kernel code and run it; for the second pass, we load the cubin file for the faster kernel directly. And multi-kernel python code is not generated for the second pass since they should not be needed.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

jansel reviewed (approved):
Please add an on-disk cache as well.  Re-autounting every run will hurt compile times.

This can be in a followup PR.

shunting314 commented:
> Please add an on-disk cache as well. Re-autounting every run will hurt compile times.

oh, sorry for making this confusing. We already have this on-disk cache enabled for multi-kernel.

shunting314 commented:
@pytorchbot merge

atalman commented:
@pytorchmergebot revert -c nosignal -m "broke linux-focal-rocm5.7-py3.8 tests"

shunting314 commented:
@atalman curious is the test 'linux-focal-rocm5.7-py3.8' somehow get skipped before I merge the PR?

shunting314 committed:
Update on "[inductor] make multi-kernel work with cpp-wrapper"


Make multi-kernel work with cpp-wrapper. multi-kernel generates two equivalent variants for a reduction. At runtime the faster one is picked. But cpp-wrapper need save cubin file during codegen. They don't work with each other at the beginning.

Thanks Jason for suggesting a neat way to integrate these two. cpp-wrapper does 2 passes codegen right now. For the first pass, we still generate multi-kernel code and run it; for the second pass, we load the cubin file for the faster kernel directly. And multi-kernel python code is not generated for the second pass since they should not be needed.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

shunting314 commented:
@pytorchbot merge

atalman commented:
@pytorchbot revert -c ghfirst -m "Failing internal tests"

shunting314 committed:
Update on "[inductor] make multi-kernel work with cpp-wrapper"


Make multi-kernel work with cpp-wrapper. multi-kernel generates two equivalent variants for a reduction. At runtime the faster one is picked. But cpp-wrapper need save cubin file during codegen. They don't work with each other at the beginning.

Thanks Jason for suggesting a neat way to integrate these two. cpp-wrapper does 2 passes codegen right now. For the first pass, we still generate multi-kernel code and run it; for the second pass, we load the cubin file for the faster kernel directly. And multi-kernel python code is not generated for the second pass since they should not be needed.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

shunting314 commented:
@pytorchbot merge

I skipped  a check in fbcode since fbcode has some issue to compile the generated cpp-wrapper cpp code.
